doc/tools/src2html1.4a/Ctags/ctags.1
.\" Ctags-new is a modified version of the ctags produced by UCB and
.\" distributed in their BSD distributions.
.\" You should be able to diff this version against theirs to see what I
.\" have changed.
.\" 	Warren Toomey
.\"
.\" Copyright (c) 1987, 1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)ctags.1	6.8 (Berkeley) 4/24/91
.\"
.Dd April 24, 1991
.Dt CTAGS 1
.Os BSD 4
.Sh NAME
.Nm ctags-new
.Nd create a tags file
.Sh SYNOPSIS
.Nm ctags-new
.Op Fl BFadtuwvx
.Op Fl f Ar tagsfile
.Ar name ...
.Sh DESCRIPTION
.Nm Ctags-new
makes a tags file for
.Xr ex 1
from the specified C,
Pascal, Fortran,
.Tn YACC ,
lex, and lisp sources.  A tags file gives the
locations of specified objects in a group of files.  Each line of the
tags file contains the object name, the file in which it is defined,
and a search pattern for the object definition, separated by white-space.
Using the
.Ar tags
file,
.Xr ex 1
can quickly locate these object
definitions.  Depending on the options provided to
.Nm ctags-new ,
objects will consist of subroutines, typedefs, defines, structs,
enums and unions.
.Bl -tag -width Ds
.It Fl B
use backward searching patterns
.Pq Li ?...? .
.It Fl F
use forward searching patterns
.Pq Li /.../
(the default).
.It Fl a
append to
.Ar tags
file.
.It Fl d
create tags for
.Li #defines
that don't take arguments;
.Li #defines
that take arguments are tagged automatically.
.It Fl f
Places the tag descriptions in a file called
.Ar tagsfile .
The default behaviour is to place them in a file
.Ar tags .
.It Fl t
create tags for typedefs, structs, unions, and enums.
.It Fl u
update the specified files in the
.Ar tags
file, that is, all
references to them are deleted, and the new values are appended to the
file.  (Beware: this option is implemented in a way which is rather
slow; it is usually faster to simply rebuild the
.Ar tags
file.)
.It Fl v
An index of the form expected by
.Xr vgrind 1
is produced on the standard output.  This listing
contains the object name, file name, and page number (assuming 64
line pages).  Since the output will be sorted into lexicographic order,
it may be desired to run the output through
.Xr sort 1 .
Sample use:
.Bd -literal -offset indent
ctags-new \-v files \&| sort \-f > index
vgrind \-x index
.Ed
.It Fl y
Yet another output format. This produces lines with the information:
symbol, line number, file name, type of symbol, each separated by whitespace.
This is used by the
.Xr src2html 1L
program.
.It Fl w
suppress warning diagnostics.
.It Fl x
.Nm ctags-new
produces a list of object
names, the line number and file name on which each is defined, as well
as the text of that line and prints this on the standard output.  This
is a simple index which can be printed out as an off-line readable
function index.
.El
.Pp
Files whose names end in
.Nm \&.c
or
.Nm \&.h
are assumed to be C
source files and are searched for C style routine and macro definitions.
Files whose names end in
.Nm \&.y
are assumed to be
.Tn YACC
source files.
Files whose names end in
.Nm \&.l
are assumed to be lisp files if their
first non-blank character is `;', `(', or `[',
otherwise, they are
treated as lex files.  Other files are first examined to see if they
contain any Pascal or Fortran routine definitions, and, if not, are
searched for C style definitions.
.Pp
The tag
.Li main
is treated specially in C programs.  The tag formed
is created by prepending
.Ar M
to the name of the file, with the
trailing
.Nm \&.c
and any leading pathname components removed.  This
makes use of
.Nm ctags-new
practical in directories with more than one
program.
.Pp
Yacc and lex files each have a special tag.
.Ar Yyparse
is the start
of the second section of the yacc file, and
.Ar yylex
is the start of
the second section of the lex file.
.Sh FILES
.Bl -tag -width tags -compact
.It Pa tags
default output tags file
.El
.Sh DIAGNOSTICS
.Nm Ctags-new
exits with a value of 1 if an error occurred, where
duplicate objects are not considered errors, 0 otherwise.
.Sh SEE ALSO
.Xr ex 1 ,
.Xr vi 1
.Sh BUGS
Recognition of
.Nm functions  ,
.Nm subroutines
and
.Nm procedures
for
.Tn FORTRAN
and Pascal is done is a very simpleminded way.  No attempt
is made to deal with block structure; if you have two Pascal procedures
in different blocks with the same name you lose.
.Nm Ctags-new
doesn't
understand about Pascal types.
.Pp
The method of deciding whether to look for C, Pascal or
.Tn FORTRAN
functions is a hack.
.Pp
.Nm Ctags-new
relies on the input being well formed, and any syntactical
errors will completely confuse it.  It also finds some legal syntax
confusing; for example, as it doesn't understand
.Li #ifdef Ns 's ,
(incidentally, that's a feature, not a bug) any code with unbalanced
braces inside
.Li #ifdef Ns 's
will cause it to become somewhat disoriented.
In a similar fashion, multiple line changes within a definition will
cause it to enter the last line of the object, rather than the first, as
the searching pattern.  The last line of multiple line
.Li typedef Ns 's
will similarly be noted.
.Sh HISTORY
The
.Nm
command appeared in 
.Bx 3.0 .


ChangeLog
2006-11-18	Ralf Corsépius <ralf.corsepius@rtems.org>

	* bootstrap: Add -p (regenerate preinstall.ams).

2006-11-15	Ralf Corsépius <ralf.corsepius@rtems.org>

	* ampolish3: New.

2006-10-17	Ralf Corsépius <ralf.corsepius@rtems.org>

	* configure.ac: Requires automake-1.10. Require autoconf-2.60.
	* aclocal/rtems-top.m4, aclocal/version.m4: 
	Adaptions to automake-1.10.

2006-09-13	Ralf Corsépius <ralf.corsepius@rtems.org>

	* bspkit: Remove.

2006-08-09	Kolja Waschk <waschk@telos.de>

	* config-ml.in: New port to Altera NIOS II.

2006-07-09      Thomas Doerfler <Thomas.Doerfler@embedded-brains.de>

	* aclocal/bsp-alias.m4: (PR1117) Add brs5l as alias for gen5200
	* aclocal/check-bsps.m4: (PR1117) Add brs5l as alias for gen5200
	
2006-06-23	Worth Burruss <worth@motioncontrol.org>

	* README.configure: Add mcf5235.

2006-04-18	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to rtems-4.6.99.3.

2006-04-18  Joel Sherrill <joel@OARcorp.com>

	* aclocal/version.m4: Updated to rtems-4.6.99.3.

2006-04-18	Joel Sherrill <joel@OARcorp.com>

	* release_support: Fix typos.

2006-04-18	Joel Sherrill <joel@OARcorp.com>

	* ChangeLog, release_support: Fix typos.

2006-03-09	Ralf Corsepius <ralf.corsepius@rtems.org>

	* acinclude.m4: Requires autoconf-2.59.

2006-03-07	Joel Sherrill <joel@OARcorp.com>

	PR 694/rtems
	* release_support: Do not include .cvsignore files in release tarballs.

2006-02-01	Ralf Corsepius <ralf.corsepius@rtems.org>

	PR 882/make_build
	* bootstrap: Only remove bspopts.h.in instead of bspopts.h*.

2006-01-12	Ralf Corsepius <ralf.corsepius@rtems.org>

	* aclocal/bsp-alias.m4: Revert Joel's 2006-01-10 change.
	Alpha-sort the aliases list.
	* automake/host.am, automake/local.am, automake/subdirs.am:
	Remove depend-am, depend-recursive.

2006-01-10	Joel Sherrill <joel@OARcorp.com>

	* aclocal/bsp-alias.m4: Add sis as alias of erc32.

2006-01-09	Ralf Corsepius <ralf.corsepius@rtems.org>

	* aclocal/bsp-alias.m4: Remove brs5l. Remove leon2.
	Alias leon1 to leon2. Add sis.
	* aclocal/check_bsps.m4: Remove brs5l.
	Remove leon. Add leon2. Add sis to erc32 family.

2005-11-31	Ralf Corsepius <ralf.corsepius@rtems.org>

	PR 851/bsps
	* aclocal/bsp-alias.m4, aclocal/check-bsps.m4: Add gen5200 BSP
	family.

2005-11-03	Ralf Corsepius <ralf.corsepius@rtems.org>

	* aclocal/check-bsps.m4: Add the plethora of mbx* BSP.

2005-09-02	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to rtems-4.6.99.2.

2005-09-02  Joel Sherrill <joel@OARcorp.com>

	* aclocal/version.m4: Updated to rtems-4.6.99.2.

2005-07-06      Philippe Simons

	* MAINTAINERS: Added me as arm/gp32 bsp maintainer.

2005-04-28	Joel Sherrill <joel@OARcorp.com>

	* README.configure: Updated to account for removal of dmv177 and
	ppcn_60x as well as other additions and deletions.

2005-02-04	Ralf Corsepius <ralf.corsepius@rtems.org>

	* aclocal/enable-itron.m4: Allow unix.

2005-01-02	Ralf Corsepius <ralf.corsepius@rtems.org>

	* Makefile.am: Remove ACLOCAL_FILES.

2004-11-16	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/check-bsps.m4: Use ls instead of globs to pickup
	bsp_specs.

2004-11-12	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/check-bsps.m4: Add MVME2100 BSP as alias of
	motorola_shared.

2004-11-10	Joel Sherrill <joel@oarcorp.com>

	* aclocal/bsp-alias.m4: Add MVME2100 BSP as alias of motorola_shared.

2004-10-22	Ralf Corsepius <ralf_corsepius@rtems.org>

	* alocal/check-custom-bsp.m4: Reflect new location of bspkit*cfg.

2004-10-18	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/check-bsps.m4: Add bspkit support. Don't check for
	presence of *.cfg.
	* aclocal/check-custom-bsp.m4: Add bspkit support.

2004-10-18	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/enable-bare.m4: Remove.

2004-09-29	Joel Sherrill <joel@OARcorp.com>

	* MAINTAINERS, README.configure, c/ACKNOWLEDGEMENTS,
	c/TOOL_TARGETS: i960 obsoleted and all references removed.

2004-09-24	Ralf Corsepius <ralf_corsepius@rtems.org>

	* configure.ac: Require automake > 1.9.

2004-09-24	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/rtems-cpu-subdirs.m4: Remove (unused).

2004-08-18	Ralf Corsepius <ralf_corsepius@rtems.org>

	PR 673/make_build
	* config-ml.in: Quote 'cd "${ml_origdir}"'.
	Quote 'cd "${ML_POPDIR}"'.

2004-08-17	Ralf Corsepius <ralf_corsepius@rtems.org>

	PR 673/make_build
	* acinclude.m4 (_RTEMS_CONFIG_SUBDIR): Quote ac_popdir uses.

2004-07-14      Joel Sherrill <joel@OARcorp.com>

        * MAINTAINERS: Added Jay Monkman as ARM architecture maintainer

2004-03-31	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/rtems-top.m4: Remove ENDIF. 
	Remove with_target_subdir defaulting to '.'.
	* acinclude.m4: Add --includedir to sub_configure_args.

2004-02-17	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* Makefile.am: DIST_SUBDIRS = $(SUBDIRS).
	Rework dist-hook.
	* configure.ac: Reflect changes to RTEMS_ENABLE_RTEMSBSP.
	* aclocal/enable-rtemsbsp.m4: Add support for --disable-rtemsbsp.

2004-02-14	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* configure.ac: RTEMS_BUILD_CONFIG_SUBDIRS([testsuites/tools]).

2004-02-13	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* Makefile.am: Remove AUTOMAKE_FILES.
	Cleanup duplicate ACLOCAL_FILES.
	EXTRA_DIST += config-ml.in.
	DIST_SUBDIRS += make.
	Add dist-hook.
	* configure.ac: Add dist-bzip2 no-dist-gzip to AUTOMAKE_OPTIONS.

2004-02-12	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* bootstrap: RTEMS_CHECK_BSPDIR(RTEMS_BSP_FAMILY).

2004-02-12	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* VERSION: 4.6.99.0.
	* aclocal/version.m4: 4.6.99.0.

2004-02-06	Joel Sherrill <joel@OARcorp.com>

	* aclocal/bsp-alias.m4: Readd aliases for p4600 and p4650
	which were accidentally removed when purging the mips64orion.

2004-02-03	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* aclocal/rpm.m4: Make osversions the default.
	* aclocal/rtems-top.m4: Remove check for version.m4 (Breaks make
	distcheck).

2004-02-01	Ralf Corsepius <corsepiu@faw.uni-ulm.de>


doc/tools/src2html1.4a/Ctags/ctags.c
/*
 * Copyright (c) 1987 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1987 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)ctags.c	5.8 (Berkeley) 2/26/91";
#endif /* not lint */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "ctags.h"

/*
 * ctags: create a tags file
 */

NODE	*head;			/* head of the sorted binary tree */

				/* boolean "func" (see init()) */
bool	_wht[0177],_etk[0177],_itk[0177],_btk[0177],_gd[0177];

FILE	*inf,			/* ioptr for current input file */
	*outf;			/* ioptr for tags file */

long	lineftell;		/* ftell after getc( inf ) == '\n' */

int	lineno,			/* line number of current line */
	dflag,			/* -d: non-macro defines */
	tflag,			/* -t: create tags for typedefs */
	wflag,			/* -w: suppress warnings */
	vflag,			/* -v: vgrind style index output */
	xflag,			/* -x: cxref style output */
	yflag;			/* -y: yet another style output */

char	*curfile,		/* current input file name */
	searchar = '/',		/* use /.../ searches by default */
	lbuf[BUFSIZ];

main(argc,argv)
	int	argc;
	char	**argv;
{
	extern char	*optarg;		/* getopt arguments */
	extern int	optind;
	static char	*outfile = "tags";	/* output file */
	int	aflag,				/* -a: append to tags */
		uflag,				/* -u: update tags */
		exit_val,			/* exit value */
		step,				/* step through args */
		ch;				/* getopts char */
	char	cmd[100];			/* too ugly to explain */

	aflag = uflag = NO;
	while ((ch = getopt(argc,argv,"BFadf:tuwvxy")) != EOF)
		switch((char)ch) {
			case 'B':
				searchar = '?';
				break;
			case 'F':
				searchar = '/';
				break;
			case 'a':
				aflag++;
				break;
			case 'd':
				dflag++;
				break;
			case 'f':
				outfile = optarg;
				break;
			case 't':
				tflag++;
				break;
			case 'u':
				uflag++;
				break;
			case 'w':
				wflag++;
				break;
			case 'v':
				vflag++;
			case 'x':
				xflag++;
				break;
			case 'y':
				yflag++;
				break;
			case '?':
			default:
				goto usage;
		}
	argv += optind;
	argc -= optind;
	if (!argc) {
usage:		puts("Usage: ctags [-BFadtuwvx] [-f tagsfile] file ...");
		exit(1);
	}

	init();

	for (exit_val = step = 0;step < argc;++step)
		if (!(inf = fopen(argv[step],"r"))) {
			perror(argv[step]);
			exit_val = 1;
		}
		else {
			curfile = argv[step];
			find_entries(argv[step]);
			(void)fclose(inf);
		}

	if (head)
		if (xflag)
			put_entries(head);
		else {
			if (uflag) {
				for (step = 0;step < argc;step++) {
					(void)sprintf(cmd,"mv %s OTAGS;fgrep -v '\t%s\t' OTAGS >%s;rm OTAGS",outfile,argv[step],outfile);
					system(cmd);
				}
				++aflag;
			}
			if (!(outf = fopen(outfile, aflag ? "a" : "w"))) {
				perror(outfile);
				exit(exit_val);
			}
			put_entries(head);
			(void)fclose(outf);
			if (uflag) {
				(void)sprintf(cmd,"sort %s -o %s",outfile,outfile);
				system(cmd);
			}
		}
	exit(exit_val);
}

/*
 * init --
 *	this routine sets up the boolean psuedo-functions which work by
 *	setting boolean flags dependent upon the corresponding character.
 *	Every char which is NOT in that string is false with respect to
 *	the pseudo-function.  Therefore, all of the array "_wht" is NO
 *	by default and then the elements subscripted by the chars in
 *	CWHITE are set to YES.  Thus, "_wht" of a char is YES if it is in
 *	the string CWHITE, else NO.
 */
init()
{
	register int	i;
	register char	*sp;

	for (i = 0; i < 0177; i++) {
		_wht[i] = _etk[i] = _itk[i] = _btk[i] = NO;
		_gd[i] = YES;
	}
#define	CWHITE	" \f\t\n"
	for (sp = CWHITE; *sp; sp++)	/* white space chars */
		_wht[*sp] = YES;
#define	CTOKEN	" \t\n\"'#()[]{}=-+%*/&|^~!<>;,.:?"
	for (sp = CTOKEN; *sp; sp++)	/* token ending chars */
		_etk[*sp] = YES;
#define	CINTOK	"ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz0123456789"
	for (sp = CINTOK; *sp; sp++)	/* valid in-token chars */
		_itk[*sp] = YES;
#define	CBEGIN	"ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz"
	for (sp = CBEGIN; *sp; sp++)	/* token starting chars */
		_btk[*sp] = YES;
#define	CNOTGD	",;"
	for (sp = CNOTGD; *sp; sp++)	/* invalid after-function chars */
		_gd[*sp] = NO;
}

/*
 * find_entries --
 *	this routine opens the specified file and calls the function
 *	which searches the file.
 */
find_entries(file)
	char	*file;
{
	register char	*cp;

	lineno = 0;				/* should be 1 ?? KB */
	if (cp = rindex(file, '.')) {
		if (cp[1] == 'l' && !cp[2]) {
			register int	c;

			for (;;) {
				if (GETC(==,EOF))
					return;
				if (!iswhite(c)) {
					rewind(inf);
					break;
				}
			}
#define	LISPCHR	";(["
/* lisp */		if (index(LISPCHR,(char)c)) {
				l_entries();
				return;
			}
/* lex */		else {
				/*
				 * we search all 3 parts of a lex file
				 * for C references.  This may be wrong.
				 */
				toss_yysec();
				(void)strcpy(lbuf,"%%$");
				pfnote("yylex",lineno);
				rewind(inf);
			}
		}
/* yacc */	else if (cp[1] == 'y' && !cp[2]) {
			/*
			 * we search only the 3rd part of a yacc file
			 * for C references.  This may be wrong.
			 */
			toss_yysec();
			(void)strcpy(lbuf,"%%$");
			pfnote("yyparse",lineno);
			y_entries();
		}
/* fortran */	else if ((cp[1] != 'c' && cp[1] != 'h') && !cp[2]) {
			if (PF_funcs())
				return;
			rewind(inf);
		}
	}
/* C */	c_entries();
}
